blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +262,238 @@ where
+
+    fn tipset_executed_message(
+        &self,
+        tipset: &Tipset,
+        cid: &Cid,
+        message: &dyn Message,
+    ) -> Result<Option<MessageReceipt>, Error> {
+        if tipset.epoch() == 0 {
+            return Ok(None);
+        }
+        let tipset = chain::tipset_from_keys(self.get_block_store_ref(), tipset.parents())
+            .map_err(|err| Error::Other(err.to_string()))?;
+        let messages = chain::message_for_tipset(self.get_block_store_ref(), &tipset)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        messages
+            .iter()
+            .enumerate()
+            .rev()
+            .filter(|(_, s)| s.from() == message.from())
+            .filter_map(|(index,s)| {
+                if s.sequence() == message.sequence() {
+                    if s.to_cid().map(|s| &s == cid).unwrap_or_default() {
+                        return Some(
+                            chain::get_parent_reciept(
+                                self.get_block_store_ref(),
+                                tipset.blocks().first().unwrap(),
+                                index as u64,
+                            )
+                            .map_err(|err| {
+                                Error::Other(err.to_string())
+                            }),
+                        );
+                    }
+                    if let Ok(m_cid) = s.to_cid()
+                    {
+                        let error_msg = format!("found message with equal nonce as the one we are looking for (F:{:} n {:}, TS: {:} n{:})",cid,message.sequence(),m_cid,s.sequence());
+                        return Some(Err(Error::Other(error_msg)))
+                    }
+                    let error_msg =format!("found message with equal nonce as the one we are looking for (F:{:} n {:}, TS: `Error Converting message to Cid` n{:})",cid,message.sequence(),s.sequence());
+                    return Some(Err(Error::Other(error_msg)))
+                }
+                if s.sequence() < message.sequence() {
+                    return Some(Ok(None));
+                }
+
+                None
+            })
+            .next()
+            .unwrap_or_else(|| Ok(None))
+    }
+
+    pub async fn wait_for_message(
+        &self,
+        cid: &Cid,
+        confidence: u64,
+    ) -> Result<Option<(Arc<Tipset>, MessageReceipt)>, Error> {
+        let subscribers = self.subscriber.clone().ok_or_else(|| {
+            Error::Other("State Manager not subscribed to tipset head changes".to_string())
+        })?;
+        let mut back_search_wait = self.back_search_wait.clone().ok_or_else(|| {
+            Error::Other("State manager not subscribed to back search wait".to_string())
+        })?;
+        let message = chain::get_chain_message(self.get_block_store_ref(), cid)
+            .map_err(|err| Error::Other(format!("failed to load message {:}", err)))?;
+
+        let subscribers: Vec<ChainMessage> = subscribers.collect().await;
```

Comment:
Shouldn't you just pull off the first item off the stream if I understand this correctly? If you collect it, you everything off that stream and can lose events.. Thoughts?

Reply:
i was really tripped out by some line and I assumed they needed a vec of events. This is true